controlplane/kubeadm/controllers/controller.go

```diff
@@ -279,6 +279,10 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 
 	controlPlane := internal.NewControlPlane(cluster, kcp, ownedMachines)
 
+	if controlPlane.HasDeletingMachine() || controlPlane.ProvisioningMachines().Len() > 0 {
```

Comment:
Yeah, the question of when to let reconciliations happen "naturally" and when to requeue is something I am not familiar with, so advice is welcome

Reply:
On a side note, IMO this test should go after `conditions.SetAggregate` so we keep updating the MachinesReadyCondition with the changes from the underlying machines

Reply:
It really depends on the configuration of watches, the predicates applied, and the mapping functions used for the specific controller. In this case in `SetupWithManager` we declare that this controller [`Owns`](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/builder#Builder.Owns) machines, which configures a predicate for the watch to reconcile any KCP resources on the creation, update, or deletion of Machines where there is an ownerref to a KCP.

Since a provisioning machine will trigger a reconciliation once the status is updated to say it has completed, we don't need to trigger an additional reconciliation to make sure that we know to reconcile again following the completion of provisioning.

For deletion, we will get updates when the deletiontimestamp is set, and any mutations (updates) of the resource, but will not get a notification when the apiserver finishes cleaning up the machine resource after the finalizers are removed, so we need to force a reconciliation to make sure that we don't need to wait up to the full resync period to continue progress.